extension/goext/database.go

```diff
@@ -67,13 +67,28 @@ func withinJoinable(tx ITransaction, fn func(tx ITransaction) error) error {
 	return fn(tx)
 }
 
+func closeTx(tx ITransaction) {
+	if tx.Closed() {
+		return
+	}
+	err := tx.Close()
+	if err != nil {
+		log.Warning(fmt.Sprintf("close scoped database transaction failed with error: %s", err))
+	}
+}
+
 func withinDetached(db IDatabase, context Context, txBegin func() (ITransaction, error), fn func(tx ITransaction) error) error {
 	opts := db.Options()
 	retryTxCount := opts.RetryTxCount
 	retryTxInterval := opts.RetryTxInterval
 	var tx ITransaction
 	var err error
 
+	defer func() {
```

Comment:
This logic is weird. IMO, we shouldn't defer `tx.Close` in root function, but right after tx creation (for loop iteration, e.g. by wrapping for loop iteration body in `func() error`) and doing defer there. Current pitfalls:
- TX is nil, because it failed to create - this code will panic on nil pointer dereference
- some TXes might remain after between retries if err checking logic is insufficient. Not this case, but it's really easy to break.

Reply:
Good points. In fact we've got a logic duplication with https://github.com/cloudwan/gohan/blob/master/db/db.go#L67 and we're fixing the same issues twice. In my opinion we should extract the common part.

Reply:
Good points. In fact we've got a logic duplication with https://github.com/cloudwan/gohan/blob/master/db/db.go#L67 and we're fixing the same issues twice. In my opinion we should extract the common part.